forge/db/controllers/ProjectSnapshot.js

```diff
@@ -1,3 +1,185 @@
+        // * device
+        // * device groups
+        // * pipeline stage device group
+        // If any of these snapshots are set as active/target, remove them from the candidates list
+
+        // Check `Devices` table
+        const query = {
+            where: {
+                [Op.or]: [
+                    { targetSnapshotId: { [Op.in]: candidateIds } },
+                    { activeSnapshotId: { [Op.in]: candidateIds } }
+                ]
+            }
+        }
+        if (typeof limit === 'number' && limit > 0) {
+            query.limit = limit
+        }
+        const snapshotsInUseInDevices = await app.db.models.Device.findAll(query)
+        const inUseAsTarget = snapshotsInUseInDevices.map((device) => device.targetSnapshotId)
+        const inUseAsActive = snapshotsInUseInDevices.map((device) => device.activeSnapshotId)
+        candidateIds = removeFromArray(candidateIds, inUseAsTarget)
+        candidateIds = removeFromArray(candidateIds, inUseAsActive)
+
+        // Check `DeviceGroups` table
+        if (app.db.models.DeviceGroup) {
+            const snapshotsInUseInDeviceGroups = await app.db.models.DeviceGroup.findAll({
+                where: {
+                    targetSnapshotId: { [Op.in]: candidateIds }
+                }
+            })
+            const inGroupAsTarget = snapshotsInUseInDeviceGroups.map((group) => group.targetSnapshotId)
+            candidateIds = removeFromArray(candidateIds, inGroupAsTarget)
+        }
+
+        // Check `PipelineStageDeviceGroups` table
+        if (app.db.models.PipelineStageDeviceGroup) {
+            const snapshotsInUseInPipelineStage = await app.db.models.PipelineStageDeviceGroup.findAll({
+                where: {
+                    targetSnapshotId: { [Op.in]: candidateIds }
+                }
+            })
+            const inPipelineStageAsTarget = snapshotsInUseInPipelineStage.map((stage) => stage.targetSnapshotId)
+            candidateIds = removeFromArray(candidateIds, inPipelineStageAsTarget)
+        }
+
+        return autoSnapshots.filter((snapshot) => candidateIds.includes(snapshot.id))
+    },
+    cleanupAutoSnapshots: async function (app, device, limit = DEVICE_AUTO_SNAPSHOT_LIMIT) {
+        // get all auto snapshots for the device (where not in use)
+        const snapshots = await app.db.controllers.ProjectSnapshot.getDeviceAutoSnapshots(device, true, 0)
+        if (snapshots.length > limit) {
+            const toDelete = snapshots.slice(0, snapshots.length - limit).map((snapshot) => snapshot.id)
+            await app.db.models.ProjectSnapshot.destroy({ where: { id: { [Op.in]: toDelete } } })
+        }
+    },
+    doAutoSnapshot: async function (app, device, deploymentType, options, meta) {
+        // sanitise options
+        options = options || {}
+        options.clean = options.clean ?? true // default to true
+        options.setAsTarget = options.setAsTarget ?? false // default to false
```

Comment:
This could also be a destructured objected in the function params

```suggestion
    doAutoSnapshot: async function (app, device, deploymentType, options = {}, meta) {
        // sanitise options
        options.clean = options.clean ?? true // default to true
        options.setAsTarget = options.setAsTarget ?? false // default to false
```